pkg/snapshot/iscsi.go

```diff
@@ -112,125 +115,136 @@ func (o *snapshotter) unmountAndDetachBlockDevice(ctx context.Context, snID stri
 	}()
 
-	// Add a portal on a target
-	out, err = exec.CommandContext(ctx, "iscsiadm", "-m", "node", "-p", defaultPortal, "-T", targetIqn, "-o", "new").CombinedOutput()
+	err = ioutil.WriteFile(path.Join(devDir, "control"), ([]byte)(fmt.Sprintf("dev_config=overlaybd//%s", confPath)), 0666)
+	if err != nil {
+		return errors.Wrapf(err, "failed to write target config for %s", devDir)
+	}
+
+	err = ioutil.WriteFile(path.Join(devDir, "enable"), ([]byte)("1"), 0666)
+	if err != nil {
+		// read the init-debug.log for readable
+		debugLogPath := o.overlaybdInitDebuglogPath(snID)
+		if data, derr := ioutil.ReadFile(debugLogPath); derr == nil {
+			return errors.Errorf("failed to enable target for %s, %s", devDir, data)
+		}
+		return errors.Wrapf(err, "failed to enable target for %s", devDir)
+	}
+
+	loopName := o.overlaybdLoopbackDeviceID(snID)
+	loopDir := fmt.Sprintf("/sys/kernel/config/target/loopback/%s", loopName)
+
+	err = os.MkdirAll(loopDir, 0700)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create loopback dir %s", loopDir)
+	}
+
+	err = os.MkdirAll(path.Join(loopDir, "tpgt_1", "lun", "lun_0"), 0700)
 	if err != nil {
-		return errors.Wrapf(err, "failed to add a portal on a target %s: %s", targetIqn, out)
+		return errors.Wrapf(err, "failed to create loopback lun dir %s", path.Join(loopDir, "tpgt_1", "lun", "lun_0"))
 	}
+
 	defer func() {
-		// rollback the portal
 		if retErr != nil {
-			deferCtx, deferCancel := rollbackContext()
-			defer deferCancel()
+			rerr := os.RemoveAll(path.Join(loopDir, "tpgt_1", "lun", "lun_0"))
+			if rerr != nil {
+				log.G(ctx).WithError(rerr).Warnf("failed to clean loopback lun %s", path.Join(loopDir, "tpgt_1", "lun", "lun_0"))
+			}
 
-			out, err = exec.CommandContext(deferCtx, "iscsiadm", "-m", "node", "-p", defaultPortal, "-T", targetIqn, "-o", "delete").CombinedOutput()
-			if err != nil {
-				log.G(deferCtx).WithError(err).Warnf("failed to rollback a portal on a target %s: %s", targetIqn, out)
+			rerr = os.RemoveAll(path.Join(loopDir, "tpgt_1"))
+			if rerr != nil {
+				log.G(ctx).WithError(rerr).Warnf("failed to clean loopback tpgt %s", path.Join(loopDir, "tpgt_1"))
+			}
+
+			rerr = os.RemoveAll(loopDir)
+			if rerr != nil {
+				log.G(ctx).WithError(rerr).Warnf("failed to clean loopback dir %s", loopDir)
 			}
 		}
 	}()
 
-	// Login a portal on a target
-	out, err = exec.CommandContext(ctx, "iscsiadm", "-m", "node", "-p", defaultPortal, "-T", targetIqn, "--login").CombinedOutput()
+	err = ioutil.WriteFile(path.Join(loopDir, "tpgt_1", "nexus"), ([]byte)(loopName), 0666)
 	if err != nil {
-		exiterr, ok := err.(*exec.ExitError)
-		if !ok || iscsi.Errno(exiterr.ExitCode()) != iscsi.ESESSEXISTS {
-			return errors.Wrapf(err, "failed to login a portal on a target %s: %s", targetIqn, out)
-		}
+		return errors.Wrapf(err, "failed to write loopback nexus %s", path.Join(loopDir, "tpgt_1", "nexus"))
+	}
+
+	err = os.Symlink(devDir, path.Join(loopDir, "tpgt_1", "lun", "lun_0", "dev_"+snID))
```

Comment:
I think we can add helper function to return the path like `path.Join(loopDir, "tpgt_1", "lun", "lun_0")` since there are too many path.Join patterns. WDYT?

Reply:
Yeah, it looks too long and complex. All paths about iSCSI and TCMU are refactored.